File: core/libs/ui/src/elements/padding.cpp
#include "elements/padding.hpp"

namespace aardvark {

float PaddingElement::get_intrinsic_height(float width) {
    return padding.height() +
           child->get_intrinsic_height(width - padding.width());
}

float PaddingElement::get_intrinsic_width(float height) {
    return padding.width() +
           child->get_intrinsic_width(height - padding.height());
}

Size PaddingElement::layout(BoxConstraints constraints) {
    auto horiz = padding.left + padding.right;
    auto vert = padding.top + padding.bottom;
    auto child_constraints = BoxConstraints{
        constraints.min_width,          // min_width
        constraints.max_width - horiz,  // max_width
        constraints.min_height,         // min_height
        constraints.max_height - vert   // max_height
    };
    auto child_size = document->layout_element(child.get(), child_constraints);
    child->size = child_size;
    child->rel_position = Position{padding.left, padding.top};
    return Size{child_size.width + horiz, child_size.height + vert};
}

}  // namespace aardvark

File: js/packages/common/src/Padding.js
const Padding = {
	all(value) {
		return { left: value, top: value, right: value, bottom: value }
	},

	only(side, value) {
		return {
			left: side === 'left' ? value : 0,
			top: side === 'top' ? value : 0,
			right: side === 'right' ? value : 0,
			bottom: side === 'bottom' ? value : 0
		}
	},

	symmetrical(horiz, vert) {
		return { left: horiz, top: vert, right: horiz, bottom: vert }
	},

	horiz(value) {
		return { left: value, top: 0, right: value, bottom: 0 }
	},

	vert(value) {
		return { left: 0, top: value, right: 0, bottom: value }
	}
}

Padding.none = Padding.all(0)

export { Padding }
